src/Upload.js

```diff
@@ -0,0 +1,158 @@
+  const scaleRatio = Math.max(scaleX, scaleY);
+  const edgeX = (scaleRatio * image.width - image.naturalWidth) / 2;
+  const edgeY = (scaleRatio * image.height - image.naturalHeight) / 2;
+  const scaledCropX = crop.x * scaleRatio;
+  const scaledCropY = crop.y * scaleRatio;
+  const scaledWidth = crop.width * scaleRatio;
+  const scaledHeight = crop.height * scaleRatio;
+  const scaledEndX = scaledCropX + scaledWidth;
+  const scaledEndY = scaledCropY + scaledHeight;
+
+  const offsetX = Math.min(Math.max(scaledCropX - edgeX, 0), image.naturalWidth);
+  const cropEndX = Math.min(Math.max(scaledEndX - edgeX, 0), image.naturalWidth);
+
+  const offsetY = Math.min(Math.max(scaledCropY - edgeY, 0), image.naturalHeight);
+  const cropEndY = Math.min(Math.max(scaledEndY - edgeY, 0), image.naturalHeight);
+
+  const cropWidthOnImage = cropEndX - offsetX;
+  const targetWidth = cropWidthOnImage / scaledWidth * width;
+  const targetOffsetX = scaledEndX >= image.naturalWidth + edgeX ? 0 : width - targetWidth;
+  const cropHeightOnImage = cropEndY - offsetY;
+  const targetHeight = cropHeightOnImage / scaledHeight * height;
+  const targetOffsetY = scaledEndY >= image.naturalHeight + edgeY ? 0 : height - targetHeight;
+
+  const canvas = document.createElement('canvas');
+
+  canvas.width = width;
+  canvas.height = height;
+  const ctx = canvas.getContext('2d');
+
+  ctx.drawImage(
+    image,
+    offsetX,
+    offsetY,
+    cropWidthOnImage,
+    cropHeightOnImage,
+    targetOffsetX,
+    targetOffsetY,
+    targetWidth,
+    targetHeight,
+  );
+
+  return new Promise((resolve, reject) => {
+    canvas.toBlob((blob) => {
+      if (!blob) {
+        reject(new Error('Canvas is empty'));
+        return;
+      }
+      resolve(blob);
+      canvas.remove();
+    }, 'image/jpeg');
+  });
+}
+
+const imageSize = { width: 40, height: 40 };
+
+export default ({ setDisplayProfilePic }) => {
+  const [src, setSrc] = useState();
+  const [imageRef, setImageRef] = useState();
+  const [blob, setBlob] = useState();
+  const [submittedToken, setSubmittedToken] = useState();
+  const { startUpload, uploads, status } = useFileUpload();
+  const [crop, setCrop] = useState({
+    unit: '%',
+    width: 30,
+    aspect: 1,
+  });
+
+  const onSelectFile = useCallback((files) => {
+    if (files && files.length > 0) {
+      const reader = new FileReader();
+      reader.addEventListener('load', () => setSrc(reader.result));
+      reader.readAsDataURL(files[0]);
+    }
+  }, [setSrc]);
+
+  const {
+    getRootProps,
+    getInputProps,
+    isDragActive,
+    isDragAccept,
+    isDragReject,
+  } = useDropzone({
+    onDrop: onSelectFile,
+    multiple: false,
+    accept: 'image/*',
+  });
+
+  useLayoutEffect(() => {
+    (async () => {
+      if (imageRef && crop.width && crop.height) {
+        const blob = await getCroppedImg(imageRef, crop, imageSize);
+        blob.name = `profile-${parseInt(Math.random() * 10000)}.jpg`;
```

Comment:
This will likely be changed by the user, so consider extracting to a documented function.  Alternatively, consider giving a unique name.

Reply:
Added a comment